middle_end/flambda/simplify/simplify_apply_expr.ml

```diff
@@ -515,12 +517,12 @@ let simplify_direct_function_call ~simplify_expr dacc apply
       let args = Apply.args apply in
       let provided_num_args = List.length args in
       let callee's_code = DE.find_code (DA.denv dacc) callee's_code_id in
-      (* A function declaration with [is_tupled = true] may effectively have
-         an arity that does not match that of the underlying code.
-         Since direct calls adopt the calling convention of the code's body
-         (whereas indirect_unknown_arity calls use the convention of the
-         function_declaration), we here always use the arity from the callee's
-         code. *)
+      (* A function declaration with [is_tupled = true] must be treated
+         specially:
+         - Direct calls adopt the normal calling convention of the code's body,
+           i.e. that given by [Code.params_arity].
+         - Indirect calls of unknown arity pass a single tuple argument,
+           irrespective of what [Code.params_arity] says. *)
```

Comment:
Maybe that comment could be tweaked a bit: I'm thinking of the case of a tupled function call that returns a function, and thus for which we could potentially have an indirect call that has more than one argument (where the first argument would be the tuple, and then other argument for the closure returned by the tupled function).

Reply:
Fixed